indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,694 @@
+    // t1 -> new batch is added to queue or batch already exists in queue
+    // t2 -> executor pops batch, processes all requests in it
+    // t1 -> new request is added to dangling batch and is never picked up
+    // Solution: For existing batch, call keyToBatch.remove() on the key to
+    // wait on keyToBatch.compute() to finish before proceeding with processBatch().
+    // For new batch, keyToBatch.remove() would not wait as key is not in map yet
+    // but a new batch is unlikely to be due immediately, so it won't get popped right away.
+    keyToBatch.compute(requestKey, (key, existingBatch) -> {
+      if (existingBatch == null) {
+        AllocateRequestBatch newBatch = new AllocateRequestBatch(key);
+        requestFuture.set(newBatch.add(request));
+        return addBatchToQueue(newBatch) ? newBatch : null;
+      } else {
+        requestFuture.set(existingBatch.add(request));
+        return existingBatch;
+      }
+    });
+
+    return requestFuture.get();
+  }
+
+  /**
+   * Tries to add the given batch to the processing queue. If the queue is full,
+   * marks the batch as completed failing all the pending requests.
+   */
+  private boolean addBatchToQueue(AllocateRequestBatch batch)
+  {
+    batch.resetQueueTime();
+    if (processingQueue.offer(batch)) {
+      log.debug("Added a new batch [%s] to queue.", batch.key);
+      return true;
+    } else {
+      log.warn("Cannot add batch [%s] as queue is full. Failing [%d] requests.", batch.key, batch.size());
+      batch.markCompleted();
+      return false;
+    }
+  }
+
+  /**
+   * Tries to add the given batch to the processing queue. If a batch already
+   * exists for this key, transfers all the requests from this batch to the
+   * existing one.
+   */
+  private void requeueBatch(AllocateRequestBatch batch)
+  {
+    log.info("Requeueing [%d] failed requests in batch [%s].", batch.size(), batch.key);
+    keyToBatch.compute(batch.key, (key, existingBatch) -> {
+      if (existingBatch == null) {
+        return addBatchToQueue(batch) ? batch : null;
+      } else {
+        // Merge requests from this batch to existing one
+        existingBatch.transferRequestsFrom(batch);
+        return existingBatch;
+      }
+    });
+  }
+
+  private void processBatchesDue()
+  {
+    // If not leader, clear the queue and do not schedule any more rounds of processing
+    if (!isLeader.get()) {
+      log.info("Not leader anymore. Failing [%d] batches in queue.", processingQueue.size());
+
+      // Keep removing items from the queue as long as not leader
+      AllocateRequestBatch nextBatch = processingQueue.peekFirst();
+      while (nextBatch != null && !isLeader.get()) {
+        processingQueue.pollFirst();
+        keyToBatch.remove(nextBatch.key);
+        nextBatch.markCompleted();
+        nextBatch = processingQueue.peekFirst();
+      }
+    }
+
+    // Check once again for leadership
+    if (!isLeader.get()) {
+      return;
+    }
```

Comment:
is this necessary? what if some items were added to the queue just before line 225?

Reply:
I guess that is not likely to happen. New items are rejected if we are not the leader.
But I guess it could happen if the leader is flapping?
- t1: we are not leader, we clear the queue
- t2: become leader
- t3: add a request
- t4: stop being leader
- t1: we are not leader, return
the request added by t3 will thus remain stuck in the queue, until we become leader again and process it. The future will timeout after 5 minutes.

I guess a simpler alternative to all of this can be to have the leadership check done for each batch separately in `processBatch`. If not leader, fail it there. The only difference would be that even non-leaders would keep polling the queue. I wanted to avoid that somehow. Please let me know what you think.